server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java

```diff
@@ -1232,6 +1232,15 @@ private static void validateCompositeTemplate(final ClusterState state,
                     for (CompressedXContent mapping : mappings) {
                         mapperService.merge(MapperService.SINGLE_MAPPING_NAME, mapping, MergeReason.INDEX_TEMPLATE);
                     }
+                    if (mapperService.documentMapper() != null
+                        && mapperService.documentMapper().routingFieldMapper() != null
+                        && mapperService.documentMapper().routingFieldMapper().required()) {
+                        if (template.getDataStreamTemplate().isAllowCustomRouting() == false) {
```

Comment:
Again, I'll have to dig into this further, but isn't this code executed for indices and not just data streams so `template.getDataStreamTemplate()` might be null?

Reply:
done